### DISCUSSION TITLE
@Transactional(readOnly = true)를 적용했을 때와 아무 어노테이션을 적용하지 않았을 때 차이

### DISCUSSION THREAD

**Discussion by gyungchan-jo** [2024-woowacourse-study/level-interview#209] [Spring Data]
.

**Comment by le2sky** [comment]
중복 디스커션입니다.
https://github.com/2024-woowacourse-study/level-interview/discussions/151

**Reply by gyungchan-jo** [reply]
공유된 디스커션은 `@Transactional` 과 `@Transactional(readOnly = true)` 의 차이를 묻는 것이라 해당 질문과 좀 다르다 생각됩니다.

**Reply by le2sky** [reply]
알겠습니다. 역따봉 취소하도록 할게요.

**Reply by gyungchan-jo** [reply]
빠른 피드백 좋네요.

**Comment by jimi567** [comment]
@Transactional(readOnly = true) 을 사용하게 된다면, DB 트랜잭션이 읽기 전용 모드로 설정 되니다.

이를 통해 다음과 같은 이점을 얻을 수 있습니다.

1. 성능 향상
 + 캐시 사용 : 읽기 전용 트랜잭션은 캐시를 활용해, DB 조회 횟수를 줄이고 어플리케이션의 응답 속도를 향상 시킬 수 있습니다.
+ DB 부하 감소 : 읽기 전용 트랜잭션의 경우 데이터의 변경이 일어나지 않기 때문에 DB Lock 전략을 가볍게 사용할 수 있습니다. 이를 통해 동시성 처리 성능을 향상 시킬 수 있습니다.

2. 데이터의 무결성 보장
 + 읽기 전용의 경우 쓰기 관련 작업이 허용되지 않기 때문에 개발자가 실수로 데이터를 변경 하려고 해도 이를 무시할 수 있습니다. 

3. 트랜잭션 간소화
+ 읽기 전용에서는 데이터 변경이 없기 때문에 트랜잭션 단위 중 예외가 발생하더라도 롤백을 할 필요가 없어서 트랜잭션 관리를 간소화 할 수 있습니다.

4. 협업 시 가독성 향상
+ 해당 메서드의 경우 읽기 작업(select)만 한다는 것을 협업하는 동료 개발자에게 쉽게 의도를 전달할 수 있게 됩니다.

**Comment by GIVEN53** [comment]
> https://github.com/woowacourse/spring-roomescape-payment/pull/24#discussion_r1621963597
서비스 레이어의 조회용 메서드에서 `@Transactional(readOnly = true)`의 유무차이로 해석
\+ repository에는 `@Transactional(readOnly = true)` 존재


```java
@Service
public class ReservationService {
    private final ReservationRepository reservationRepository;

    public ReservationService(ReservationRepository reservationRepository) {
        this.reservationRepository = reservationRepository;
    }

    public String getMemberName(Long id) {
        Reservation reservation = reservationRepository.findById(id).orElseThrow(...);
        return reservation.getMember().getName(); // LazyInitializationException
    }
}
```
OSIV 옵션이 false라고 가정했을 때 영속성 컨텍스트는 트랜잭션이 종료될 때 닫히게 되어 영속성 컨텍스트에 관리되고 있는 엔티티는 준영속 상태가 됩니다.
이 때 Lazy 로딩이 설정된 연관관계 엔티티를 조회하면 LazyInitializationException이 발생합니다.
따라서 Lazy 로딩이 필요할 경우 `@Transactional`이 필요합니다.

> readOnly 옵션의 유무와는 크게 상관없는 내용인 것 같아서 서비스 레이어에 `@Transactional`의 유무에 따른 조회 차이로 해석하는 것도 나쁘지 않을 듯 합니다

<br/>

### 번외
> 질문 해석 잘못해서 코드 깠다가 아까워서 냅둠

`@Transactional(readOnly = true)`을 사용할 경우 읽기 전용 모드가 적용되며 flush 모드를 MANUAL로 설정합니다.
> 명시적으로 flush를 호출하지 않으면 flush가 자동으로 수행되지 않는 모드
> <img width="682" alt="스크린샷 2024-06-24 오전 12 49 00" src="https://github.com/2024-woowacourse-study/level-interview/assets/101033262/98d805a4-0a6b-4d96-b4db-0134ec6e7fb1">

따라서 트랜잭션 내에서 강제로 `flush()`를 호출하지 않는 한 dirty checking이 반영되지 않습니다.

_JpaTransactionManager_.doBegin() -> _HibernateJpaDialect_.beginTransaction() -> _HibernateJpaDialect_.prepareFlushMode()
<img width="723" alt="스크린샷 2024-06-24 오전 12 54 08" src="https://github.com/2024-woowacourse-study/level-interview/assets/101033262/60a7010e-c5bf-4e25-8c2a-2713b747e567">
